### ISSUE TITLE
System.Range package similar to System.ValueTuple

### ISSUE THREAD

**Issue by lostmsu** [dotnet/corefx#34273]
My understanding is `System.Range` and `System.Index` are going to be .NET Core 3.0 specific features.
That prevents me from creating a library, that targets .NET Standard 2.0 (for I need to support .NET 4.7), and implements range indexers for the types defined in it.

I think it is a huge problem, especially considering the next .NET Standard is not going to be implemented by .NET 4.8.

Are there any plans to deal with it?

Is the problem in the desire to implement range-based indexing for the core types like `System.Array`? If so, are there ways to work around that? Can .NET Core 3.0 instead define `System._Range`, and have hypothetical System.Range package implementing an implicit conversion `System._Range` -> `System.Range`? Any other alternatives?

**Comment by gfoidl** [comment]
> Any other alternatives?

You can use [multi-targeting](https://weblog.west-wind.com/posts/2017/Jun/22/MultiTargeting-and-Porting-a-NET-Library-to-NET-Core-20). So you have a code path for .NET Core 3.0 and one for .NET Standard 2.0.

**Comment by gfoidl** [comment]
I meant the latter, and don't know a better way.

**Comment by joshfree** [comment]
@tarekgh can you PTAL

**Comment by tarekgh** [comment]
@lostmsu Index and Range are not in netstandard 2.0. These types will be in netstandard 2.1. You may look at https://github.com/dotnet/standard/pull/1018.

**Comment by lostmsu** [comment]
@tarekgh , yes, and that is the nature of the problem. The indexing feature C# 8 will be unusable on .NET Framework.

**Comment by eerhardt** [comment]
It's not just the indexing feature of C# 8 being available on .NET Framework that I'm concerned about. (There are other C# 8 features that flat out won't ever work on .NET Framework 4.8. See [Building C# 8.0 Blog](https://devblogs.microsoft.com/dotnet/building-c-8-0/)).

But the core issue here is that you can't have a `netstandard2.0` or less library that exposes APIs that will allow the indexing feature in C# 8 to work when it IS available (say in a .NET Core 3.0 app).

For example, in the `System.Numerics.Tensors` library (which targets `netstanard1.1` today), we'd like to have a `Slice` method on a `Tensor`. Since tensors are multi-dimensional, you need to have a collection of `start` and `stop` indices - basically exactly what `System.Range` is. One for each dimension.

To expose the `Slice` API, we need some way of specifying the range in each dimension you want to slice off the tensor. Without having `System.Range` available, `Numerics.Tensors` will need to come up with their own "Range-like API".  Let's say `ValueTuple<int, int>`. But here's the problem - is the 2nd int a Length or a stopping index? If it is a stopping index, is it inclusive or exclusive? Does it handle negative indices? etc. etc.

All of these problems would be solved if we had a common type that encapsulated these semantics. And we do have the type, we just aren't allowing libraries less than `netstandard2.1` use it.

**Comment by johnnyasantoss** [comment]
Any updates on a package for `Range` and `Index`?

**Comment by bgrainger** [comment]
In lieu of an official solution, I created my own unofficial one last month:

* https://github.com/bgrainger/IndexRange
* https://www.nuget.org/packages/IndexRange/

**Comment by tarekgh** [comment]
@terrajobst @jaredpar do we have the final call on that?
